sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -110,6 +110,31 @@ trait CheckAnalysis {
               s"filter expression '${f.condition.sql}' " +
                 s"of type ${f.condition.dataType.simpleString} is not a boolean.")
 
+          case f @ Filter(condition, child) =>
+            // Make sure no correlated predicate is in an OUTER join, because this could change the
+            // semantics of the join.
+            lazy val attributes: Set[Expression] = child.output.toSet
+            def checkCorrelatedPredicates(p: PredicateSubquery): Unit = p.query.foreach {
+              case j @ Join(left, right, jt, _) if jt != Inner =>
+                j.transformAllExpressions {
+                  case e if attributes.contains(e) =>
+                    failAnalysis(s"Accessing outer query column is not allowed in outer joins: $e")
```

Comment:
Then `l.id` should be part of right child of join, not part of the join condition, right?